src/core/Logger.mli

```diff
@@ -0,0 +1,46 @@
+module type Handler = DiagnosticEmitter.Handler
+
+module type S =
+sig
+  module Code : Code.S
+  module Diagnostic : Diagnostic.S with module Code := Code
+
+  module type Handler = Handler with module Code := Code and module Diagnostic := Diagnostic
+
+  (** [messagef ~loc ~additional_marks ~code format ...] constructs a diagnostic along with the backtrace frames recorded via [tracef]. *)
+  val messagef : ?loc:Span.t -> ?additional_marks:Span.t list -> code:Code.t -> ('a, Format.formatter, unit, Diagnostic.t) format4 -> 'a
+
+  (** [kmessagef kont ~loc ~additional_marks ~code format ...] constructs a diagnostic and then apply [kont] to the resulting diagnostic. *)
+  val kmessagef : (Diagnostic.t -> 'b) -> ?loc:Span.t -> ?additional_marks:Span.t list -> code:Code.t -> ('a, Format.formatter, unit, 'b) format4 -> 'a
+
+  (** [tracef ~loc format ...] record a frame. *)
+  val tracef : ?loc:Span.t -> ('a, Format.formatter, unit, (unit -> 'b) -> 'b) format4 -> 'a
+
+  (** [append_marks msg marks] appends [marks] to the additional marks of [msg]. *)
+  val append_marks : Diagnostic.t -> Span.t list -> Diagnostic.t
```

Comment:
Perhaps this should be called `with_marks`?

Reply:
For this I'm less sure. I feel `with_marks marks msg` could be interpreted in at least three ways:

1. Replace the additional marks in `msg` with `marks`.
2. Prepend `marks` to the additional marks in `msg`.
3. Append `marks` to the additional marks in `msg`.

The current name `append_marks` made it clear that it's 3. Any suggestions to address the ambiguity?

Reply:
Good point! In that case `append_marks` is fine. The reason I suggested the change was to make
```
messagef ~code:SomeCode "some error"
|> append_marks some_marks
|> Perform.fatal
```

a little nicer to read, but this is fine! (and we may as well just use the `additional_marks` argument anyways)